lib/maze/browser_stack_utils.rb

```diff
@@ -9,16 +11,21 @@ class << self
       # @param username [String] the BrowserStack username
       # @param access_key [String] the BrowserStack access key
       def upload_app(username, access_key, app)
-        # TODO: Improve error handling:
-        #   - res may not be JSON at all
         if app.start_with? 'bs://'
           app_url = app
           $logger.info "Using pre-uploaded app from #{app}"
         else
-          url = 'https://api-cloud.browserstack.com/app-automate/upload'
-          res = `curl -u "#{username}:#{access_key}" -X POST "#{url}" -F "file=@#{app}"`
-          response = JSON.parse(res)
-          raise "BrowserStack upload failed due to error: #{response['error']}" if response.include?('error')
+          $logger.info "Uploading app: #{app}"
+          res = RestClient.post "https://#{username}:#{access_key}@api-cloud.browserstack.com/app-automate/upload",
```

Comment:
I'm a bit wary of using rest-client as it doesn't look maintained anymore. This can be done with the standard library in only a few more lines, something along the lines of:

```ruby
uri = URI("https://#{username}:#{access_key}@api-cloud.browserstack.com/app-automate/upload")
request = Net::HTTP::Post.new(uri)
request.set_form({ File.basename(app) => File.new(app, 'rb') }, 'multipart/form-data')

res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
  http.request(request)
end
```

Reply:
I did try using Ruby's Net::HTTP first, but was not able to get it working with BrowserStack's API.

The suggested code (slightly amended for basic auth) results in a `Broken pipe (Errno::EPIPE)` error, which is tried using rest-client (which works out of the box.)

```ruby
uri = URI("https://api-cloud.browserstack.com/app-automate/upload")
request = Net::HTTP::Post.new(uri)
request.basic_auth username, access_key
request.set_form({ File.basename(app) => File.new(app, 'rb') }, 'multipart/form-data')
res = Net::HTTP.start(uri.hostname, uri.port) do |http|
  http.request(request)
end
```